Add excel import and export to littletable

bvanelli opened the pull request:
I stumbled upon your project and thought I could add something new. Since a lot of data actually comes from excel sheets, I thougth an excel importer/exporter would be great. Here is an example input (included on the test folder):

![image](https://user-images.githubusercontent.com/8211602/146101525-8d7cf2aa-b26b-4efe-a387-3135d12c0a37.png)

Here is an example import code:

```python
from littletable import Table

excel_table = Table("excel")

excel_table.excel_import(
    "test/abc.xlsx", transforms={"a": int, "b": int, "c": int}, limit=5
)
print(excel_table.as_markdown())

excel_table.excel_export("test_modified.xlsx", fieldnames="a bra cadabra")
```

Here is the output:

![image](https://user-images.githubusercontent.com/8211602/146101672-88d1f251-0ff9-4f6e-b617-37272099da20.png)

I could not fully run the tests offline because of the http server not working as expected, but looks like they are working as far as I could tell.

Regarding code style, I left to intepretation since there is nothing mentioned on the docs.

Best regards,
Brunno

ptmcg commented:
Looks neat! I'll give it a try this weekend.

ptmcg commented:
Tests all pass for me, not sure why they are failing in your environment.

I also see that data types like int and float values get preserved when imported, which is nice.

I can merge it as-is - there are a few changes I'd like to make about the import of openpyxl, and maybe raising a friendlier exception if it is not installed, similar to how rich is handled. But I'm glad to see the tests, and it will be a nice enhancement, thanks for submitting!

ptmcg commented:
I ran some more tests this morning, with various data types. It works better than I expected! openpyxl does a nice job detecting and converting float, int, str, bool, and even datetime data types.

ptmcg commented:
Did a little more studying of openpyxl. For a performance boost during import and export, openpyxl has read_only and write_only modes.

Consider adding this patched code:

```
Index: littletable.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/littletable.py b/littletable.py
--- a/littletable.py	(revision e244dd7f8e4e12c4f966aecb622c2b50fba6c0a8)
+++ b/littletable.py	(date 1640012543101)
@@ -2269,21 +2269,21 @@
             # lazy import openpyxl
             from openpyxl import load_workbook
 
-            wb = load_workbook(filename)
-            # read requested sheet if provided on kwargs, otherwise read active sheet
-            requested_sheet = reader_args.get("sheet")
-            ws = wb[requested_sheet] if requested_sheet else wb.active
-            # check whether to include or omit the header
-            header = reader_args.get("fieldnames") or [cell.value for cell in ws[1]]
-            start_position = 0 if reader_args.get("fieldnames") else 1
-
-            # return read data as List[Dict]
-            data = list()
-            for row in list(ws.rows)[start_position:]:
-                values = {}
-                for key, cell in zip(header, row):
-                    values[key] = cell.value
-                data.append(values)
+            with closing(load_workbook(filename, read_only=True)) as wb:
+                # read requested sheet if provided on kwargs, otherwise read active sheet
+                requested_sheet = reader_args.get("sheet")
+                ws = wb[requested_sheet] if requested_sheet else wb.active
+                # check whether to include or omit the header
+                header = reader_args.get("fieldnames") or [cell.value for cell in ws[1]]
+                start_position = 0 if reader_args.get("fieldnames") else 1
+    
+                # return read data as List[Dict]
+                data = list()
+                for row in list(ws.rows)[start_position:]:
+                    values = {}
+                    for key, cell in zip(header, row):
+                        values[key] = cell.value
+                    data.append(values)
             return data
 
         return self._import(
@@ -2524,8 +2524,16 @@
         **kwargs,
     ):
         from openpyxl import Workbook
-        wb = Workbook()
-        ws = wb.active
+        try:
+            import lxml
+        except ImportError:
+            lxml = None
+        if lxml is not None:
+            wb = Workbook(write_only=True)
+            ws = wb.create_sheet()
+        else:
+            wb = Workbook()
+            ws = wb.active
         # set header rows
         if fieldnames is None:
             fieldnames = self._attr_names()

```

Or if you prefer, I'll merge your PR as-is, and then make some post-merge edits.

bvanelli commented:
Hi, I added your patch to the mr to auto close the files and performance improvements, with the little change of allowing lxml to be disabled (I had some problems with it before, especially when using pyinstaller. It's something very very specific but it's worth leaving the option open).

I also added the same exception handling for openpyxl import as the others.

The problem with the tests is because it uses `subprocess.Popen`, which calls an internal system command and depends on the system path, and might not work if you have multiple python versions or weird path mappings (i.e. python maps to python3 instead of python).

bvanelli commented:
Btw, his fixes the tests for me (running in a thread instead):

It's a bigger change since I had to rename test to tests because of internal python conflict.

```patch
diff --git a/test/abc.csv b/tests/abc.csv
similarity index 100%
rename from test/abc.csv
rename to tests/abc.csv
diff --git a/test/abc.csv.gz b/tests/abc.csv.gz
similarity index 100%
rename from test/abc.csv.gz
rename to tests/abc.csv.gz
diff --git a/test/abc.csv.xz b/tests/abc.csv.xz
similarity index 100%
rename from test/abc.csv.xz
rename to tests/abc.csv.xz
diff --git a/test/abc.csv.zip b/tests/abc.csv.zip
similarity index 100%
rename from test/abc.csv.zip
rename to tests/abc.csv.zip
diff --git a/test/abc.json.gz b/tests/abc.json.gz
similarity index 100%
rename from test/abc.json.gz
rename to tests/abc.json.gz
diff --git a/test/abc.tsv b/tests/abc.tsv
similarity index 100%
rename from test/abc.tsv
rename to tests/abc.tsv
diff --git a/test/csv_import_http_server.py b/tests/csv_import_http_server.py
similarity index 64%
rename from test/csv_import_http_server.py
rename to tests/csv_import_http_server.py
index 7d2ef21..2e201b0 100644
--- a/test/csv_import_http_server.py
+++ b/tests/csv_import_http_server.py
@@ -1,6 +1,9 @@
 from http.server import HTTPServer, BaseHTTPRequestHandler
 from http import HTTPStatus
 import threading
+import urllib.request
+import urllib.error
+import time


 # noinspection PyPep8Naming
@@ -12,6 +15,9 @@ class CSVTestRequestHandler(BaseHTTPRequestHandler):
             self.send_response(HTTPStatus.OK)
             self.end_headers()
             threading.Thread(target=lambda: self.server.shutdown()).start()
+        elif path == "/":
+            self.send_response(HTTPStatus.OK)
+            self.end_headers()

         if path.startswith("/abc.csv"):
             send_bytes = b"a,b,c\n1,2,3\n"
@@ -26,5 +32,18 @@ def run(server_class=HTTPServer, handler_class=CSVTestRequestHandler):
     httpd.serve_forever()


+def run_background_test_server():
+    p = threading.Thread(target=run)
+    p.start()
+    # try to get the endpoint until it works
+    for _ in range(20):
+        try:
+            with urllib.request.urlopen("http://localhost:8888/"):
+                break
+        except urllib.error.URLError:
+            time.sleep(0.1)
+    return p
+
+
 if __name__ == '__main__':
     run()
diff --git a/unit_tests.py b/unit_tests.py
index 1b9fa6c..9606eab 100644
--- a/unit_tests.py
+++ b/unit_tests.py
@@ -15,6 +15,7 @@ import textwrap
 from types import SimpleNamespace
 import unittest
 from typing import Optional, Union, NamedTuple
+from tests import csv_import_http_server

 import littletable as lt

@@ -1556,7 +1557,7 @@ class TableImportExportTests:
         self.assertEqual(type(t1[0]), type(csvtable2[0]))

     def test_csv_compressed_import(self):
-        tt = lt.Table().csv_import("test/abc.csv", transforms=dict.fromkeys("abc", int))
+        tt = lt.Table().csv_import("tests/abc.csv", transforms=dict.fromkeys("abc", int))
         print("abc.csv", tt.info())

         compressed_files = [
@@ -1565,7 +1566,7 @@ class TableImportExportTests:
             "abc.csv.xz",
         ]
         for name in compressed_files:
-            import_source_name = "test/" + name
+            import_source_name = "tests/" + name
             tt2 = lt.Table().csv_import(import_source_name, transforms=dict.fromkeys("abc", int))
             print(name, tt2.info())
             expected_info = {**tt.info(), "name": import_source_name}
@@ -1575,8 +1576,8 @@ class TableImportExportTests:
             self.assertEqual(sum(tt.all.c), sum(tt2.all.c))

         # test separately, no transforms for JSON imports
-        import_source_name = "test/abc.json.gz"
-        tt2 = lt.Table().json_import("test/abc.json.gz")
+        import_source_name = "tests/abc.json.gz"
+        tt2 = lt.Table().json_import("tests/abc.json.gz")
         print("abc.json.gz", tt2.info())
         expected_info = {**tt.info(), "name": import_source_name}
         self.assertEqual(expected_info, tt2.info())
@@ -1595,7 +1596,7 @@ class TableImportExportTests:
         ]
         for fname, expected_type in imports:
             if "\n" not in fname:
-                import_name = "test/" + fname
+                import_name = "tests/" + fname
             else:
                 import_name = fname
             if not import_name.endswith(".tsv"):
@@ -1612,11 +1613,12 @@ class TableImportExportTests:
             self.assertEqual(expected_type, tbl.import_source_type)

     def test_csv_import_from_url(self):
-        import subprocess
         import urllib.request

         web_address = "http://localhost:8888"
-        web_server = subprocess.Popen("python test/csv_import_http_server.py".split())
+
+        p = csv_import_http_server.run_background_test_server()
+
         url = web_address + "/abc.csv"
         tbl = lt.Table().csv_import(url)
         tbl.present()
@@ -1626,24 +1628,24 @@ class TableImportExportTests:
         with urllib.request.urlopen(web_address + "/EXIT"):
             pass

-        web_server.wait()
+        p.join()

     def test_csv_filtered_import(self):
         test_size = 3
-        tt = lt.Table().csv_import("test/abc.csv", transforms=dict.fromkeys("abc", int))
+        tt = lt.Table().csv_import("tests/abc.csv", transforms=dict.fromkeys("abc", int))
         print("abc.csv", tt.info())

-        tt = lt.Table().csv_import("test/abc.csv", transforms=dict.fromkeys("abc", int),
+        tt = lt.Table().csv_import("tests/abc.csv", transforms=dict.fromkeys("abc", int),
                                    filters={"c": lt.Table.eq(1)})
         print(tt.info())
         self.assertEqual(test_size * test_size, len(tt))

-        tt = lt.Table().csv_import("test/abc.csv", transforms=dict.fromkeys("abc", int),
+        tt = lt.Table().csv_import("tests/abc.csv", transforms=dict.fromkeys("abc", int),
                                    filters={"c": 1})
         print(tt.info())
         self.assertEqual(test_size * test_size, len(tt))

-        tt = lt.Table().csv_import("test/abc.csv", transforms=dict.fromkeys("abc", int),
+        tt = lt.Table().csv_import("tests/abc.csv", transforms=dict.fromkeys("abc", int),
                                    filters={"c": lambda x: 0 < x < 2})
         print(tt.info())
         self.assertEqual(test_size * test_size, len(tt))
```

ptmcg commented on littletable.py:
```diff
@@ -2252,6 +2261,94 @@ def tsv_import(
+        if openpyxl is None:
+            raise Exception("openpyxl module not installed")
+
+        def excel_as_dict(filename, **reader_args) -> List[Dict[str, str]]:
+            with closing(openpyxl.load_workbook(filename, read_only=True)) as wb:
+                # read requested sheet if provided on kwargs, otherwise read active sheet
+                requested_sheet = reader_args.get("sheet")
+                ws = wb[requested_sheet] if requested_sheet else wb.active
+                # check whether to include or omit the header
+                header = reader_args.get("fieldnames") or [cell.value for cell in ws[1]]
+                start_position = 0 if reader_args.get("fieldnames") else 1
+
+                # return read data as List[Dict]
+                data = list()
+                for row in list(ws.rows)[start_position:]:
+                    values = {}
+                    for key, cell in zip(header, row):
+                        values[key] = cell.value
+                    data.append(values)
+            return data
```
Another stab at excel_as_dict, this time using iterators and making it a generator, so that we minimize the number of copies of the data.

        def excel_as_dict(filename, **reader_args) -> Iterable[Dict[str, Any]]:
            # lazy import openpyxl
            from openpyxl import load_workbook

            with closing(load_workbook(filename, read_only=True)) as wb:
                # read requested sheet if provided on kwargs, otherwise read active sheet
                requested_sheet = reader_args.get("sheet")
                ws = wb[requested_sheet] if requested_sheet else wb.active

                rows_iter = iter(ws.rows)

                # check whether to include or omit the header
                header = (reader_args.get("fieldnames")
                          or [cell.value for cell in next(rows_iter)])

                for row in rows_iter:
                    yield {key: cell.value for key, cell in zip(header, row)}

bvanelli replied:
Did the changes in https://github.com/ptmcg/littletable/pull/4/commits/58eb3f9a11d7e6672a2bc7295186b09558460f86

ptmcg commented:
Looks good to merge this PR now. We can work on future changes in another PR. (I had difficulty doing an excel_import from a URL - might want to give that a look.)

I'll look at the Popen -> threading change separately, but what if you used sys.executable to get the current running Python, and use that instead of "python" in the command? Like this:

    def test_csv_import_from_url(self):
        import subprocess
        import urllib.request

        python_exe = sys.executable
        cmd = [python_exe, "test/csv_import_http_server.py"]
        web_address = "http://localhost:8888"
        url = web_address + "/abc.csv"
        try:
            web_server = subprocess.Popen(cmd)
            tbl = lt.Table().csv_import(url)
        finally:
            with urllib.request.urlopen(web_address + "/EXIT"):
                pass

            web_server.wait()

        tbl.present()
        self.assertEqual(url, tbl.import_source)
        self.assertEqual(lt.ImportSourceType.url, tbl.import_source_type)
